adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -735,41 +946,210 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+        if (f._1.isEmpty) {
+          // use the bounds from the left partition map
+          destinationPartitionMap(f._2.toInt)
+        } else {
+          f._1
+        }
+      }).collect
+
+    replaceRdd(finalPartitionedRDD.values, Some(newPartitionMap))
   }
+
+  /**
+   * Gets the partition bounds from a ReferenceRegion keyed Iterator
+   *
+   * @param iter The data on a given partition. ReferenceRegion keyed
+   * @return The bounds of the ReferenceRegions on that partition, in an Iterator
+   */
+  private def getRegionBoundsFromPartition(iter: Iterator[(ReferenceRegion, T)]): Iterator[Option[(ReferenceRegion, ReferenceRegion)]] = {
+    if (iter.isEmpty) {
+      // This means that there is no data on the partition, so we have no bounds
+      Iterator(None)
+    } else {
+      val firstRegion = iter.next
+      val lastRegion =
+        if (iter.hasNext) {
+          // we have to make sure we get the full bounds of this partition, this
+          // includes any extremely long regions. we include the firstRegion for
+          // the case that the first region is extremely long
+          (iter ++ Iterator(firstRegion)).maxBy(f => (f._1.referenceName, f._1.end, f._1.start))
+          // only one record on this partition, so this is the extent of the bounds
+        } else {
+          firstRegion
+        }
+      Iterator(Some((firstRegion._1, lastRegion._1)))
+    }
+  }
+
+  /**
+   * Sort and repartition the data. We use Sparks sortByKey on the
+   * ReferenceRegions of a given record. This gets us data that is
+   * partitioned to be roughly uniform in size across all nodes.
+   *
+   * @param partitions The number of partitions to split the data into
+   * @return A U that contains the sorted and partitioned RDD
+   */
+  def repartitionAndSort(partitions: Int = rdd.partitions.length)(implicit c: ClassTag[T]): U = {
+    // here we let spark do the heavy lifting
+    val partitionedRDD = flattenRddByRegions()
+      .sortBy(f => f._1, ascending = true, partitions)
+    partitionedRDD.cache()
+    val tempPartitionMap = partitionedRDD.mapPartitions(iter =>
+      getRegionBoundsFromPartition(iter), preservesPartitioning = true).collect
+
+    // we want to try to avoid any None values in our final partition map
+    val newPartitionMap = tempPartitionMap.zipWithIndex.map(f => {
+      if (f._1.isEmpty && f._2 != 0) {
+        // in order to prevent None's, we just use the previous element's
+        // upper bound to build this partition bound.
+        Some({
+          val previousBound = tempPartitionMap(f._2 - 1).get._2
+          (ReferenceRegion(previousBound.referenceName, previousBound.start + 1, previousBound.end + 1),
+            ReferenceRegion(previousBound.referenceName, previousBound.start + 2, previousBound.end + 2))
+
+        })
+      } else {
+        f._1
+      }
+    })
+
+    replaceRdd(partitionedRDD.values, Some(newPartitionMap))
+  }
+
+  private case class PreparedForShuffleRegionJoin[T, X](
```

Comment:
I agree to eliminate the class. Now that our parameter needs for `ShuffleRegionJoin` are so low, we don't need a case class.